### PR TITLE
Storage migration tests

### PR DESCRIPTION
**Pull request by maxsam4**
`migration-tests/remote-externalities/` has been borrowed from substrate and modified to work with our codebase. Rest of it is new code.

You can run the tests via
```
BUILD_DUMMY_WASM_BINARY=1 cargo test -p migration-tests --lib -- --nocapture
```

### CODE REVIEW

File: migration-tests/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+/// This is the main helper function for writing migration tests,
+/// This function takes two closures as input. It first executes the first closure on current blockchain state.
+/// It then executes the storage migrations and finally it executes the second closure.
+pub fn test_migration<F, G>(pre_tests: F, post_tests: G)
+where
+    F: FnOnce() -> (),
+    G: FnOnce() -> (),
+{
```
Comment (Centril) [inline migration-tests/src/lib.rs:41]:
```suggestion
/// Main helper function for writing migration tests.
///
/// Takes two closures, executing the first on the current blockchain state,
/// then the storage migrations, and finally the second closure.
pub fn test_migration(pre_tests: impl FnOnce(), post_tests: impl FnOnce()) {
```

**Review by Centril — APPROVED** [review]

File: migration-tests/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+struct TestState;
+
+impl TestState {
+    fn init() {
+        if !std::path::Path::new(".").join("CACHE").exists() {
+            block_on(
+                Builder::new()
+                    .mode(Mode::Online(OnlineConfig {
+                        uri: "http://159.69.94.51:9933".into(),
```
Comment (satyamakgec) [inline migration-tests/src/lib.rs:25]: Do we want this URL to open? isn't it better to have some environment variables?

**Review by satyamakgec — COMMENTED** [review]
Could it be possible to move the `remove-externalities` into the fork of ours so It will be cleaner?

File: migration-tests/src/lib.rs (hunk 2 above)
Comment (satyamakgec) [reply migration-tests/src/lib.rs:25]: A chance of DDoS

Comment (maxsam4) [reply migration-tests/src/lib.rs:25]: Think of it as a honey pot. I'd rather have the dumb script kiddies dos this than our other infra :)

**Comment by maxsam4** [comment]
@satyamakgec `remove-externalities` is not part of substrate right now.

**Review by satyamakgec — APPROVED** [review]

**Comment by maxsam4** [comment]
Do not delete the branch. Migration tests depend on this.

**Pull request merged**